### PR TITLE
CORE-15820 - Token Selection - Add new version of the token state observer api

### PR DESCRIPTION
**Pull request by filipesoliveira**
Updated the token selection code to support the new API version of the token state observer.

This PR depends on: https://github.com/corda/corda-api/pull/1228

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -46,7 +46,7 @@ commonsTextVersion = 1.10.0
 bouncycastleVersion=1.73
 # Corda API libs revision (change in 4th digit indicates a breaking change)
 # Change to 5.0.0.xx-SNAPSHOT to pick up maven local published copy
-cordaApiVersion=5.1.0.19-beta+
+cordaApiVersion=5.1.0.19-SNAPSHOT
```
Comment (filipesoliveira) [inline gradle.properties]: This change must be reverted before merging the code!

File: testing/cpbs/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/UtxoDemoTokenStateObserver.kt [hunk 2]
```diff
@@ -12,8 +11,8 @@ const val TOKEN_SYMBOL = "USD"
 const val TOKEN_TYPE = "TestUtxoState"
 val TOKEN_AMOUNT = BigDecimal.TEN
 
-@Suppress("UNUSED")
```
Comment (owenstanford) [inline testing/cpbs/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/UtxoDemoTokenStateObserver.kt:15]: I think we can remove the demo classes if they are not used in the smoke tests

File: testing/cpbs/packaging-verification-contract-v1/src/main/kotlin/com/r3/corda/testing/packagingverification/contract/SimpleTokenStateObserver.kt [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
 
 import net.corda.v5.application.crypto.DigestService
-import net.corda.v5.ledger.utxo.observer.UtxoLedgerTokenStateObserver
 import net.corda.v5.ledger.utxo.observer.UtxoToken
 import net.corda.v5.ledger.utxo.observer.UtxoTokenFilterFields
 import net.corda.v5.ledger.utxo.observer.UtxoTokenPoolKey
 
-class SimpleTokenStateObserver : UtxoLedgerTokenStateObserver<SimpleState> {
+@Suppress("DEPRECATION")
+class SimpleTokenStateObserver : net.corda.v5.ledger.utxo.observer.UtxoLedgerTokenStateObserver<SimpleState> {
```
Comment (owenstanford) [inline testing/cpbs/packaging-verification-contract-v1/src/main/kotlin/com/r3/corda/testing/packagingverification/contract/SimpleTokenStateObserver.kt]: I think we should change the smoke tests to use the new observer type

File: testing/cpbs/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/UtxoDemoTokenStateObserver.kt (hunk 2 above)
Comment (filipesoliveira) [reply testing/cpbs/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/UtxoDemoTokenStateObserver.kt:15]: This observer is still used by the tests.

File: gradle.properties [hunk 4]
```diff
@@ -47,7 +47,7 @@ bouncycastleVersion=1.73
 # Corda API libs revision (change in 4th digit indicates a breaking change)
 # Change to 5.0.0.xx-SNAPSHOT to pick up maven local published copy
 #cordaApiVersion=5.1.0.xxx-SNAPSHOT
-cordaApiVersion=5.1.0.19-beta+
+cordaApiVersion=5.1.0.20-alpha-1694776169307
```
Comment (filipesoliveira) [inline gradle.properties]: This must be updated before merging.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/request/handlers/UtxoPersistTransactionRequestHandler.kt [hunk 5]
```diff
@@ -32,44 +33,74 @@ class UtxoPersistTransactionRequestHandler @Suppress("LongParameterList") constr
...
+            // Look for an observer that implements the new interface
+            val observerV2 = tokenObservers.getObserverForV2(stateAndRef.state.contractStateType)
+            if (observerV2 != null) {
+                return@flatMap  onCommit(observerV2, stateAndRef) { obs, sAndRef ->
+                    obs.onCommit(
+                        sAndRef,
+                        transactionReader.getUtxoTransaction(persistenceService),
+                        digestService
+                    )
+                }
```
Comment (blsemo) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/request/handlers/UtxoPersistTransactionRequestHandler.kt]: Shouldn't that be the other way round, i.e. calling the new interface first? I assume that over time, there will be many more apps using the new interface than the old one...

Comment (filipesoliveira) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/request/handlers/UtxoPersistTransactionRequestHandler.kt]: Done.

**Comment by filipesoliveira** [comment]
The following comment must be addressed before merging the code:
https://github.com/corda/corda-runtime-os/pull/4531#discussion_r1331799348

Done.

**Review by owenstanford — DISMISSED** [review]

File: gradle.properties [hunk 6]
```diff
@@ -46,7 +46,7 @@ commonsTextVersion = 1.10.0
 bouncycastleVersion=1.73
 # Corda API libs revision (change in 4th digit indicates a breaking change)
 # Change to 5.0.0.xx-SNAPSHOT to pick up maven local published copy
-cordaApiVersion=5.1.0.21-beta+
+cordaApiVersion=5.1.0.22-alpha-1695372427878
```
Comment (filipesoliveira) [inline gradle.properties]: This must be updated before merging the code.

**Review by kyriathar — APPROVED** [review]
OK (db + packaging changes)

**Review by owenstanford — APPROVED** [review]

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/request/handlers/UtxoPersistTransactionRequestHandler.kt [hunk 7]
```diff
@@ -54,22 +48,57 @@ class UtxoPersistTransactionRequestHandler @Suppress("LongParameterList") constr
...
 
-    private fun List<StateAndRef<ContractState>>.toTokens(tokenObservers: UtxoTokenObserverMap): List<Pair<StateAndRef<*>, UtxoToken>> =
-        flatMap { stateAndRef ->
+    private fun getOutputTokenRecords(
+        listOfPairsStateAndUtxoToken: List<Pair<StateAndRef<*>, UtxoToken>>
+    ): List<Record<TokenPoolCacheKey, TokenPoolCacheEvent>> {
+        val isTransactionVerified = transaction.status == TransactionStatus.VERIFIED
+        if (!isTransactionVerified) {
+            return listOf()
+        }
```
Comment (blsemo) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/request/handlers/UtxoPersistTransactionRequestHandler.kt:57]: That local variable is not needed.

**Review by blsemo — APPROVED** [review]
Looks ok - the code could be clearer in one place by removing a local variable, but fine to merge as is.

**Pull request merged**